cmd/noms/diff/diff.go

```diff
@@ -5,121 +5,138 @@
+// - see Diff vs DiffLeftRight in Set and Map.
+//
+// Note: the function sends messages on diffChan and checks whether stopChan has
+// been closed to know if it needs to terminate diffing early. To function
+// properly it needs to be executed concurrently with code that reads values from
+// diffChan. The following is a typical invocation of Diff():
+//    dChan := make(chan Difference)
+//    sChan := make(chan struct{})
+//    go func() {
+//        d.Diff(s3, s4, dChan, sChan, leftRight)
+//        close(dChan)
+//    }()
+//    for dif := range dChan {
+//        <some code>
+//    }
+func Diff(v1, v2 types.Value, dChan chan<- Difference, stopChan chan struct{}, leftRight bool) {
+	d := differ{diffChan: dChan, stopChan: stopChan, leftRight: leftRight}
+	if !v1.Equals(v2) {
+		if !shouldDescend(v1, v2) {
+			d.sendDiff(Difference{Path: nil, ChangeType: types.DiffChangeModified, OldValue: v1, NewValue: v2})
+		} else {
+			d.diff(nil, v1, v2)
+		}
 	}
-
-	return diff(dw, types.Path{}, nil, v1, v2)
 }
 
-func diff(w diffWriter, p types.Path, key, v1, v2 types.Value) error {
+func (d differ) diff(p types.Path, v1, v2 types.Value) bool {
 	switch v1.Type().Kind() {
 	case types.ListKind:
-		return diffLists(w, p, v1.(types.List), v2.(types.List))
+		return d.diffLists(p, v1.(types.List), v2.(types.List))
 	case types.MapKind:
-		return diffMaps(w, p, v1.(types.Map), v2.(types.Map))
+		return d.diffMaps(p, v1.(types.Map), v2.(types.Map))
 	case types.SetKind:
-		return diffSets(w, p, v1.(types.Set), v2.(types.Set))
+		return d.diffSets(p, v1.(types.Set), v2.(types.Set))
 	case types.StructKind:
-		return diffStructs(w, p, v1.(types.Struct), v2.(types.Struct))
+		return d.diffStructs(p, v1.(types.Struct), v2.(types.Struct))
 	default:
 		panic("Unrecognized type in diff function")
 	}
 }
 
-func diffLists(w diffWriter, p types.Path, v1, v2 types.List) (err error) {
+func (d differ) diffLists(p types.Path, v1, v2 types.List) (stop bool) {
 	spliceChan := make(chan types.Splice)
-	stopChan := make(chan struct{}, 1) // buffer size of 1, so this won't block if diff already finished
+	stopChan := make(chan struct{}, 1) // buffer size of 1s, so this won't block if diff already finished
 
 	go func() {
 		v2.Diff(v1, spliceChan, stopChan)
 		close(spliceChan)
 	}()
 
-	wroteHdr := false
-
 	for splice := range spliceChan {
-		if err != nil {
+		if stop {
 			break
 		}
-
 		if splice.SpRemoved == splice.SpAdded {
 			// Heuristic: list only has modifications.
-			for i := uint64(0); i < splice.SpRemoved && err == nil; i++ {
+			for i := uint64(0); i < splice.SpRemoved; i++ {
 				lastEl := v1.Get(splice.SpAt + i)
 				newEl := v2.Get(splice.SpFrom + i)
 				if shouldDescend(lastEl, newEl) {
 					idx := types.Number(splice.SpAt + i)
-					writeFooter(w, &wroteHdr)
-					err = diff(w, append(p, types.NewIndexPath(idx)), idx, lastEl, newEl)
+					stop = d.diff(append(p, types.NewIndexPath(idx)), lastEl, newEl)
```

Comment:
You need to copy this too.

I agree with your observation that this is a reallocation, maybe instead of `append(copyPath(p), foo)` it chould be `copyAppend(p, foo)` then you can allocate a buffer of size +1.